File: src/main/java/ui_adapter/model/ActionResult.java
package ui_adapter.model;

import ui_adapter.error.ErrorType;

public class ActionResult {
    public enum Status {
        PASS,
        FAIL
    }

    private final Status status;
    private final String message;
    private final String screenshotPath;
    private final ErrorType errorType;
    private final String pageState;  // Added: captures page content/state for validation

    public ActionResult(Status status, String message, String screenshotPath, ErrorType errorType) {
        this(status, message, screenshotPath, errorType, null);
    }

    public ActionResult(Status status, String message, String screenshotPath, ErrorType errorType, String pageState) {
        this.status = status;
        this.message = message;
        this.screenshotPath = screenshotPath;
        this.errorType = errorType;
        this.pageState = pageState;
    }

    public static ActionResult pass(String message) {
        return new ActionResult(Status.PASS, message, null, null, null);
    }

    public static ActionResult pass(String message, String pageState) {
        return new ActionResult(Status.PASS, message, null, null, pageState);
    }

    public static ActionResult fail(String message, ErrorType errorType, String screenshotPath) {
        return new ActionResult(Status.FAIL, message, screenshotPath, errorType, null);
    }

    public static ActionResult fail(String message, ErrorType errorType, String screenshotPath, String pageState) {
        return new ActionResult(Status.FAIL, message, screenshotPath, errorType, pageState);
    }

    public Status getStatus() {
        return status;
    }

    public String getMessage() {
        return message;
    }

    public String getScreenshotPath() {
        return screenshotPath;
    }

    public ErrorType getErrorType() {
        return errorType;
    }

    public String getPageState() {
        return pageState;
    }

    @Override
    public String toString() {
        return "ActionResult{" +
                "status=" + status +
                ", message='" + message + '\'' +
                ", screenshotPath='" + screenshotPath + '\'' +
                ", errorType=" + errorType +
                ", pageState=" + (pageState != null ? "<captured>" : "<none>") +
                '}';
    }
}

File: python-code-base/playwright_ui_adapter/model/action_result.py
"""Action result model representing the outcome of action execution."""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Optional

from ..error.error_types import ErrorType


class ActionStatus(str, Enum):
    """Action execution status."""
    
    PASS = "PASS"
    FAIL = "FAIL"
    
    def __str__(self) -> str:
        return self.value


@dataclass
class ActionResult:
    """
    Result of executing a UI action.
    
    Contains status, message, error details, and optionally captured
    page state for validation and debugging.
    
    Attributes:
        status: Execution status (PASS/FAIL)
        message: Human-readable result message
        error_type: Optional error categorization for failures
        screenshot_path: Optional path to failure screenshot
        page_state: Optional captured page content for validation
        timestamp: When the result was created
    """
    
    status: ActionStatus
    message: str
    error_type: Optional[ErrorType] = None
    screenshot_path: Optional[str] = None
    page_state: Optional[str] = None
    timestamp: datetime = field(default_factory=datetime.now)
    
    @classmethod
    def success(cls, message: str, page_state: Optional[str] = None) -> "ActionResult":
        """
        Factory method for successful result.
        
        Args:
            message: Success message
            page_state: Optional captured page state
            
        Returns:
            ActionResult with PASS status
        """
        return cls(
            status=ActionStatus.PASS,
            message=message,
            page_state=page_state
        )
    
    @classmethod
    def failure(
        cls,
        message: str,
        error_type: ErrorType,
        screenshot_path: Optional[str] = None,
        page_state: Optional[str] = None
    ) -> "ActionResult":
        """
        Factory method for failed result.
        
        Args:
            message: Failure message
            error_type: Error categorization
            screenshot_path: Optional path to failure screenshot
            page_state: Optional captured page state
            
        Returns:
            ActionResult with FAIL status
        """
        return cls(
            status=ActionStatus.FAIL,
            message=message,
            error_type=error_type,
            screenshot_path=screenshot_path,
            page_state=page_state
        )
    
    def __str__(self) -> str:
        parts = [f"{self.status}: {self.message}"]
        if self.error_type:
            parts.append(f"[{self.error_type}]")
        if self.screenshot_path:
            parts.append(f"screenshot={self.screenshot_path}")
        return " ".join(parts)
    
    def __repr__(self) -> str:
        return f"ActionResult({self.status}, {self.message})"
